### PR TITLE
KAFKA-2527; System Test for Quotas in Ducktape

### PR DESCRIPTION
**Pull request by lindong28**
@granders Can you take a look at this quota system test?

### CODE REVIEW

File: tests/kafkatest/tests/quota_test.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            self.logger.info("consumer %s has maximum throughput %.2f bps, average throughput %.2f bps with quota %.2f bps" %
+                             (client_id, maximum_bps, self.consumer_average_bps[client_id], quota_bps))
+            if maximum_bps > quota_bps*(self.maximum_deviation_percentage/100+1):
+                success = False
+                msg += "maximum consumer throughput %.2f bps exceeded quota %.2f bps by more than %.1f%%" % \
+                       (maximum_bps, quota_bps, self.maximum_deviation_percentage)
+
+        return success, msg
+
+    def getProducerQuota(self, client_id):
```
Comment (granders) [inline tests/kafkatest/tests/quota_test.py]: use snake-case (underscores) rather than camel-case

File: tests/kafkatest/services/performance/jmx_mixin.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kafkatest.services.performance import PerformanceService
+
+
+class JmxMixin(PerformanceService):
```
Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: I don't think `JmxMixin` should subclass `PerformanceService`, for a couple reasons:
- Some classes that might want to use this logic won't be performance services
- The idea with mixins is to make use of multiple inheritance and provide reusable functionality

In other words, this class should subclass `object`

File: tests/kafkatest/services/performance/producer_performance.py [hunk 3]
```diff
@@ -13,10 +13,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from kafkatest.services.performance import PerformanceService
+from kafkatest.services.performance.jmx_mixin import JmxMixin
 
-
-class ProducerPerformanceService(PerformanceService):
+class ProducerPerformanceService(JmxMixin):
```
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: Inherit from both `PerformanceService` and `JmxMixin` (see comment on mixin above)
i.e.
`class ProducerPerformanceService(PerformanceService, JmxMixin)`

File: tests/kafkatest/services/performance/jmx_mixin.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        cmd = "/opt/kafka/bin/kafka-run-class.sh kafka.tools.JmxTool " \
+              "--reporting-interval 1000 --jmx-url service:jmx:rmi:///jndi/rmi://127.0.0.1:%d/jmxrmi " \
+              "--object-name %s " % (self.jmx_port, self.jmx_object_name)
+        if self.jmx_attributes != None:
+            cmd += "--attributes %s " % self.jmx_attributes
+        cmd += "> /mnt/jmx_tool.log &"
+
+        self.logger.debug("Start JmxTool %d command: %s", idx, cmd)
+        for l in node.account.ssh_capture(cmd, allow_fail=False):
```
Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: since you're not capturing output, use `node.account.ssh(cmd ...)`

File: tests/kafkatest/services/performance/jmx_mixin.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        node.account.ssh("rm -rf /mnt/jmx_tool.log", allow_fail=False)
+
+    def maybe_start_jmx_tool(self, idx, node):
+        if self.started[idx-1] == True or self.jmx_object_name == None:
+            return
+        self.started[idx-1] = True
+
+        cmd = "/opt/kafka/bin/kafka-run-class.sh kafka.tools.JmxTool " \
+              "--reporting-interval 1000 --jmx-url service:jmx:rmi:///jndi/rmi://127.0.0.1:%d/jmxrmi " \
+              "--object-name %s " % (self.jmx_port, self.jmx_object_name)
```
Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: JmxTool allows multiple jmx object names - should this do the same?

Also, we may want to support running the tool with no object-name specified (so that it queries all objects)

File: tests/kafkatest/services/performance/producer_performance.py [hunk 6]
```diff
@@ -62,6 +63,7 @@ def parse_stats(line):
             }
         last = None
         for line in node.account.ssh_capture(cmd):
+            self.maybe_start_jmx_tool(idx, node)
```
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: Calling `maybe_start_jmx_tool` after each line that gets read from the producer process doesn't seem quite right. 

I think we want the behavior to be:
- start producer "asynchronously"
- start jmx tool asynchronously
- wait for producer to finish
- process each line of producer output

I think it would look something like:

```
cmd = "same_as_before &"  # now the cmd is "async"
node.account.ssh(cmd)
wait_until(producer is alive)

self.start_jmx_tool(node)

wait_until(producer is finished)
for line in node.account.ssh_capture("cat /mnt/producer-performance.log"):
  # same as the previous for loop
```

File: tests/kafkatest/tests/quota_test.py (hunk 1 above)
Comment (lindong28) [inline tests/kafkatest/tests/quota_test.py]: Thanks for taking time to review this patch. Will fix this one.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 2 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Sure. Will fix it.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 3 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/producer_performance.py]: Will fix it.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 4 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Even though I don't need output, if I don't wait for output, then the cmd will not be executed by ssh_capture. Does it make sense? I will test it again and get back to you.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 5 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Yes, JmxTool allows multiple object names and no object name as argument. But should we make it full-fledged from the beginning?

I currently implemented in such a way to only satisfy the use case where we query exactly one attribute. Can we expand it later when we have use case for multiple object names or attributes?

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/producer_performance.py]: Yes I have thought of the approach as you described -- I didn't implement it because it required the test case developer to specifically redirect output to a file and read from that file, which should not be required for developer to do if we can the right support from Ducktape. This is the reason I asked for asynchronous ssh in Ducktape during our discussion in the email. 

I am thinking of something like this:

```
cmd = "same_as_before"
output = node.account.ssh_asynchrous(cmd)
self.start_jmx_tool(node)
for line in output:
   # same as the previous for loop
```

Does it make sense?

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 4 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: I just double checked it by running the test case -- simply do `node.account.ssh_capture(cmd, allow_fail=False)` without waiting for output Ducktape will not execute the cmd. I couldn't explain this though -- most likely this is related to how python execute the function with `yield`.

Can we fix it in Ducktape side so that `node.account.ssh(cmd ...)` always execute the cmd? Are you able to reproduce the problem I found?

Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Maybe I'm misunderstanding - did you try `node.account.ssh` or `node.account.ssh_capture`, or both?

The first simply runs the command; the second tries to capture output.

Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Ah.. my bad. Problem fixed after using `node.account.ssh`. Thanks for the suggestion.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 5 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Ah, I realized there is a better interface for JmxMixin to provide min, avg, and max value for every queried object-attribute. I will update the patch to support this. Thanks for the suggestion.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (ewencp) [inline tests/kafkatest/services/performance/producer_performance.py]: @lindong28 Now it's a lot clearer what you meant in your original email. You're basically saying that since `RemoteAccount.ssh()` blocks, `RemoteAccount.ssh_output()` blocks until all output is collected and `RemoteAccount.ssh_capture()` drives the whole process despite acting as a generator, there's no way to start a second process between when you start the first and when you start reading its output?

I think it makes a lot of sense to provide support for that. More generally, I think the current set of `RemoteAccount.ssh*` methods could use some cleanup and potentially renaming to simplify things. Originally I thought the blocking `RemoteAccount.ssh()` would be the most widely used, but it turns out we've found more and more uses for capturing the output.

@granders, a patch just to add Ducktape support for the approach @lindong28 is looking for should be quick and easy. I think it can look a lot like `ssh_capture`, but it'll have to return an iterable object after invoking the ssh command rather than treating the entire method call as a generator using yield. It might even make sense to just return the subprocess object and allow the user to do whatever they want with it, so the example would look something like

```
cmd = "same_as_before"
proc = node.account.ssh_async(cmd)
self.start_jmx_tool(node)
for line in iter(proc.stdout.readline, ''):
   # same as the previous for loop
```

I think there are cleaner ways to expose this, but given how many `ssh*` variants we're ending up with, we may want to hold off on introducing much more until we sort out how to capture all of them with a minimum of methods.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 5 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: @granders JmxMixin has been updated to support query with arbitrary object name and attributes. Thanks for the suggestion.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/producer_performance.py]: @ewencp Thanks for your comment. As you mentioned in the ticket, I can file a follow up patch after Ducktape is updated.

**Comment by lindong28** [comment]
@granders @ewencp @guozhangwang I have addressed comments received so far and all 3 quota test succeed on my desktop.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/producer_performance.py]: @ewencp In response to your question, yeah that is what I have in mind when I was thinking about the test, and is the reason I wanted to have support for "asynchronous" ssh command from Ducktape. But I didn't stated it very clearly in the email, partly because something become clearer only after I started implementation -- for example, I once thought that adding "&" would solve problem cleanly but it didn't.

**Comment by granders** [comment]
@lindong28 fyi I am seeing a failure in test_shared_quota when I run on aws:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/90/console

Summary report:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/90/artifact/results/KAFKA-2527/2015-10-06--001/report.html

Artifacts:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/90/artifact/results/KAFKA-2527/2015-10-06--001/

**Comment by lindong28** [comment]
@granders Thanks for your information. This also happens occasionally in my test. This test try to verify that, when 2 consumer shares the same quota, maximum value of sum of their throughput at any given time <= quota \* (1+50%). However, it sometimes exceeds the quota by almost 100%. I need to investigate 1) the quota implementation in kafka and 2) metrics calculation in the kafka and client source code to understand why deviation can be as much as 100%. 

As far as quota system test is concerned, I have modified the threshold to 100% to suppress the problem.

**Comment by granders** [comment]
@lindong28 
Doesn't upping the threshold to suppress the failure defeat the purpose of this particular test?

I would think it'd be better to have a test that fails + a followup JIRA to investigate whether the failure is real than a false pass.

**Comment by ewencp** [comment]
Agreed it defeats the purpose of the test. Didn't something like this come up at some point during the evaluation of the original patch, maybe in one of the unit tests or something? I think one proposed possibility was that the windows used to compute stats for the quota system and for metrics was different, which could account for the difference.

File: tests/kafkatest/tests/quota_test.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    def setUp(self):
+        self.zk.start()
+        self.kafka.start()
+
+    def min_cluster_size(self):
+        """Override this since we're adding services outside of the constructor"""
+        return super(QuotaTest, self).min_cluster_size() + self.num_producers + self.num_consumers
+
+    def run_clients(self, producer_id, producer_num, consumer_id, consumer_num):
```
Comment (granders) [inline tests/kafkatest/tests/quota_test.py:76]: fyi - this is essentially a parametrized test, so you could express it that way:

change "run_clients" -> "test_quotas", and then:

```
from ducktape.marks import parameterize

...

@parameterize(producer_id="default_id", producer_num=1, consumer_id="default_id", consumer_num=1)
@parameterize(producer_id="overridden_id", producer_num=1, consumer_id="overridden_id", consumer_num=1)
@parameterize(producer_id="overridden_id", producer_num=1, consumer_id="overridden_id", consumer_num=2)
def test_quotas(self, producer_id, producer_num, consumer_id, consumer_num):
    stuff...

```

**Comment by lindong28** [comment]
@granders @ewencp I think the purpose of system tests is to notify developers of bugs so we can fix it. But it is not clear that there is any bug in the quota implementation even though the deviation can be high, for the following reasons:

-- Immediate goal of quote is to ensure that the maximum throughput per clientId on broker averaged over a given window <= quota. This is verified in the original evaluation when quota >= 10MB (posted in KAFKA-2528). But it doesn't guarantee that the maximum throughput measured on the client side <= quota, because 1) there is variable delay between broker and consumer, and 2) consumer uses different configuration and implementation for its rate metrics.

-- This high deviation happens in the shared quota test, which may additionally be explained by the variable delay between broker and consumers. For example, say broker allocates 100% quota to consumer 1 at time t, and allocates 100% quota to consumer 2 at time t+1. If the batches of messages arrives at both consumers at the same time, then the sum of measured byte-in-rate would double the quota.

Overall the quota is working, because without quota you would observe much higher byte-in rate or byte-out rate in the test. I would say there is problem if the sum of average throughput over entire experiment >= quota. I plan to do further investigation to explain the gap here. But we probably should not mark test as FAIL if this doesn't indicate problem in kafka implementation.

**Comment by lindong28** [comment]
Thinking more about this, maybe I should also collect byte-in rate and byte-out rate directly from broker in this system test.

File: tests/kafkatest/tests/quota_test.py (hunk 7 above)
Comment (ewencp) [inline tests/kafkatest/tests/quota_test.py:76]: It'd also be nice to use default values on the method parameters so the `@parametrize` annotations only have to specify the parameters that are overridden.

**Comment by ewencp** [comment]
@lindong28 Yeah, I think that explanation of how you can get a window where the client sees the quota exceeded was the explanation I had seen, probably in that discussion in KAFKA-2528. I think the key thing in this test is to figure out the exact ways in which the measured value can exceed the quota and document that clearly in the test. I think the 2x number makes sense -- as long as we never exceed that then we can consider it valid. Measuring byte-in and byte-out rates directly would also be good for validation. Those shouldn't skew as much, but could also deviate since they are measured at a different point than the quota is enforced, right? What would be the acceptable deviation from the quota there?

File: tests/kafkatest/tests/quota_test.py (hunk 7 above)
Comment (lindong28) [inline tests/kafkatest/tests/quota_test.py:76]: @granders I haven't seen any use of this feature in kafkatests. Does this allow developer to specify test names, e.g. test_overridden_quota, so that we can get an idea of what the test is in the report?

Comment (ewencp) [inline tests/kafkatest/tests/quota_test.py:76]: @lindong28 In the console output the tests get a name with the parameters included, e.g. `kafkatest.tests.benchmark_test.Benchmark.test_producer_throughput.topic=topic-replication-factor-one.num_producers=1.acks=1.message_size=100`. In the HTML report, the parameters are listed with the test name in the "Test" column. See, e.g., the report [http://testing.confluent.io/kafka/latest/](here) and search for the `test_producer_throughput` tests.

If you're looking for examples, we use this in a few places in the tests already -- the copycat and benchmark tests are both using it.

**Comment by lindong28** [comment]
@granders @ewencp I think the purpose of system tests is to notify developers of bugs so we can fix it. But it is not clear that there is any bug in the quota implementation even though the deviation can be high, for the following reasons:

-- Immediate goal of quote is to ensure that the maximum throughput per clientId on broker averaged over a given window <= quota. This is verified in the original evaluation when quota >= 10MB (posted in KAFKA-2528). But it doesn't guarantee that the maximum throughput measured on the client side <= quota, because 1) there is variable delay between broker and consumer, and 2) consumer uses different configuration and implementation for its rate metrics.

-- The deviation in the single client is much

**Comment by lindong28** [comment]
Ah.. Closed by mistake.

File: tests/kafkatest/tests/quota_test.py (hunk 7 above)
Comment (lindong28) [inline tests/kafkatest/tests/quota_test.py:76]: @ewencp @granders Very nice feature. Thanks for sharing. I thought it is not used in kafkatest because I searched for "parameterize" instead of "parametrize". For consistency I will use "parametrize" as well.

The only inconvenience with that approach is that I can not selectively run one of the three test (no?). But this is more of a problem during development and debug.

I find it clearer to specify every parameter in the python code than specifying difference set of parameters in the "@parametrize". Since this is also the practice adopted in benchmark_test and copycat_test, I will specify every parameter as well.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: @lindong28 Thanks for this observation. I added a fix in https://github.com/confluentinc/ducktape/pull/107

The fix is now available from pypi as version 0.3.5 - you'll want to update the ducktape dependency in tests/setup.py

Note this is actually asynchronous even without running command as a background process with "&", so you should be able to run:

```
producer_output = node.account.ssh_capture(cmd)
wait_until(producer is alive)
self.start_jmx_tool(node)

for line in producer_output:
  process(line)
```

File: tests/kafkatest/services/performance/jmx_mixin.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                continue
+            stats = [float(field) for field in line.split(',')]
+            time_sec = int(stats[0]/1000)
+            self.jmx_stats[idx-1][time_sec] = {name : stats[i+1] for i, name in enumerate(object_attribute_names)}
+
+        # do not calculate average and maximum of jmx stats until we have read output from all nodes
+        if any(len(dict)==0 for dict in self.jmx_stats):
+            return
+
+        start_time_sec = min([min(dict.keys()) for dict in self.jmx_stats])
```
Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Rename "dict" here since it is a reserved keyword

File: tests/kafkatest/services/performance/jmx_mixin.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        cmd = "/opt/kafka/bin/kafka-run-class.sh kafka.tools.JmxTool " \
+              "--reporting-interval 1000 --jmx-url service:jmx:rmi:///jndi/rmi://127.0.0.1:%d/jmxrmi" % self.jmx_port
+        if self.jmx_object_name != None:
+            cmd += " --object-name %s" % self.jmx_object_name
+        if self.jmx_attributes != None:
+            cmd += " --attributes %s" % self.jmx_attributes
+        cmd += " > /mnt/jmx_tool.log &"
+
+        self.logger.debug("Start JmxTool %d command: %s", idx, cmd)
+        node.account.ssh(cmd, allow_fail=False)
```
Comment (granders) [inline tests/kafkatest/services/performance/jmx_mixin.py]: We've found it's very helpful with debugging test failures to provide some sort of fail-fast guarantee that the process we care about is up and running before returning from these start commands

E.g.

```
def start_jmx_tool(...):
   node.account.ssh(cmd...) 
   wait_until(jmx_tool_is_alive)
   # now it's ok to return
```

Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Sure. I will fix it.

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 8 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: Thanks for noting this. Will fix it.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 6 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/producer_performance.py]: Thanks @granders for the patch. Will use the new API.

**Comment by lindong28** [comment]
@ewencp In response to your earlier comments on the acceptable deviation -- in theory, the broker's byte-in rate or byte-out rate could deviate from quota's rate sensor for reasons you mentioned. I wouldn't be able to comment on the exact acceptable deviation percentage except for doing the test. Fortunately it appears that the maximum per-minute broker byte-in rate and byte-out rate are always lower than their respective quota. Please see below evaluation results when 2 consumers share the quota.

Based on the following results I would bound the deviation between broker's byte rate and quota by 5%. Does it sound good?

```
producer produced 100000 messages
consumer consumed 100000 messages
producer has maximum throughput 4487708.66 bps with producer quota 3750000.00 bps
broker has maximum byte-in rate 2792396.09 bps with producer quota 3750000.00 bps
consumer has maximum throughput 4397383.20 bps with consumer quota 3000000.00 bps
broker has maximum byte-out rate 2527560.97 bps with consumer quota 3000000.00 bps
```

**Comment by granders** [comment]
@lindong28 fyi I just merged a ducktape PR (https://github.com/confluentinc/ducktape/pull/108) which makes it possible to specify test parameters at the command-line (version `0.3.6`). Hopefully this makes parametrization more convenient.

E.g. 
`ducktape tests/.../a_test.py::TestClass.test_something --parameters '{"x": 1, "y": "a_string"}'`

**Comment by ewencp** [comment]
@lindong28 It does look like the broker rates are under the quotas, but it'd be good to reason from first principles what the right limit is so we'll see a regression if it is violated. If the broker rates are per-minute and the quotas are computed per second, wouldn't the maximum limit be `(61 * quota_rate / 60)`? We always just need to match up the measured period with the period of the rate limiting, with some slack to account for mismatches in time windows.

**Comment by lindong28** [comment]
@ewencp  BrokerTopicMetrics uses EWMA, configured to be ticked every 5 seconds, to calculate 1-minute rate. Quota uses the sliding window algorithm, configured with 11 samples of 1 second each, to calculate 11-second rate. Since 60-seconds can be covered by 11-seconds_6, a possibly loose bound is that `max(broker_metric_rate) <= max(quota_rate)_66/60 = (1+10%)*max(quota_metric_rate)`

However, note that there is no bound on how much `quota_metric_rate` can exceed `quota` since the quota metric is updated before we decide whether to throttled request. The deviation depends on the number of inflight requests between broker and clients. Adding to the complexity, there is also difference in when the broker's byte-in rate and quota's byte-in rate are updated -- broker's byte-in rate is updated immediate after broker appends data to log, while quota's byte-in rate is updated after necessary acks are received from follower.

In short, 10% seems a reasonable bound on the deviation between broker's rate and quota. Though there is no guarantee that this will be satisfied all the time, from users's perspective it is good enough if in most tests the `maximum broker rate <= (1+10%) quota` .

File: tests/kafkatest/services/performance/jmx_mixin.py (hunk 9 above)
Comment (lindong28) [inline tests/kafkatest/services/performance/jmx_mixin.py]: @granders Do you think generator returned from ssh can support hasNext()? This can be useful to wait for the process to run, instead of having an extra remote ssh command to tail the log file, as is done by kafka service.

The code would look like this:

```
def start_jmx_tool(...):
   output = node.account.ssh_capture(cmd...) 
   output.hasNext()
   # now it's ok to return
```

**Comment by lindong28** [comment]
@ewencp @granders @guozhangwang I have updated patch to address the comments received so far. Thanks for your help.

File: tests/kafkatest/services/console_consumer.py [hunk 10]
```diff
@@ -169,18 +172,24 @@ def _worker(self, idx, node):
...
 
+        self.read_jmx_output(idx, node)
+
     def start_node(self, node):
-        super(ConsoleConsumer, self).start_node(node)
+        PerformanceService.start_node(self, node)
 
     def stop_node(self, node):
-        node.account.kill_process("java", allow_fail=True)
+        node.account.kill_process("console-consumer", allow_fail=True)
```
Comment (granders) [inline tests/kafkatest/services/console_consumer.py]: There is now one sanity check which fails both remotely and on my mac:
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/94/console

This is fixed by changing `console-consumer` to `console_consumer` (dash to underscore)

**Comment by granders** [comment]
@lindong28 
Thanks a lot for your adjustments!

In hindsight, I think jmxtool possibly has too much internal state to be a good mixin, and the multiple inheritance here is evidently making it a bit hard to use - e.g. there seem to be a few awkward and dangerous things happening with a mix of old and new-style calls to superclass methods which I think were sortof necessitated by the multiple inheritance and the method signatures.

That said, if you adjust the `stop_node` call so `test_console_consumer` passes, this seems basically good.

A couple of things should happen in a followup
- incorporate the ducktape update which allows the async `ssh_capture` call and remove the `ssh_capture(..).next()` trick which can cause problems if there are no lines of output (`next()` raises `StopIteration` error if there is no output, as in the `test_console_consumer` sanity check)
  - I can tweak the jmx tool and the old-style super calls when I merge your changes into my upcoming patch

**Comment by lindong28** [comment]
@granders Thanks much as well for your review and help to update Ducktape! The `stop_node` is fixed now.

I have two questions:

-- Could you elaborate a bit more on why there is too much internal state in JmxMixin? It seems to me that this JmxMixin will have no performance overhead if user doesn't want to query any jmx attribute. And it requires only the necessary overhead, e.g. start jmx process at beginning and read and analyze jmx log at the end, to provide user with average and maximum value for each jmx attribute. In other words, is there any better way to allow multiple service query jmx attribute via mixin?

-- I have used `ssh_capture(..).next()` in order to wait for producer/consumer to start before I start jmx tool. I think the best solution is for ducktape to support `ssh_capture(..).hasNext()`. Do you think this is a useful API for Ducktape to provide? If developer were to wait for process to start without this support, they have to do `ps ax | grep -i console_consumer` (in ConsoleConsumer) or `tail kafka.log` (in KafkaService), both require extra ssh_command.

I agree with you that the way clients (e.g. producer performance and console consumer) calls super class methods is awkward due to multiple inheritance and method signatures. It will be nice if we can find a way to solve this problem.

**Comment by lindong28** [comment]
@guozhangwang @granders @ewencp Can you help review this patch so that it can hopefully get committed by Oct 16 as requested by Jun for 0.9.0 release?

**Comment by granders** [comment]
@lindong28 
- Let's address any jmx cleanup in a followup patch.
- `has_next()` or some other mechanism to signal that output is available would be useful, maybe you can open an issue on ducktape for a feature request? https://github.com/confluentinc/ducktape
- The initial SSL test patch was just merged, so there are some conflicts to be resolved, but @gwenshap can take a final look once that is ready

**Comment by lindong28** [comment]
@granders @gwenshap Sure, I just merged my patch with trunk. Quota tests have passed on my machine. Please have a look. Thanks!

**Comment by lindong28** [comment]
@granders Sure, I added a feature ticket in https://github.com/confluentinc/ducktape/issues/109. Can I submit a patch for this feature?

**Comment by granders** [comment]
@lindong28 Yes, patches welcome!

**Comment by granders** [comment]
@lindong28 
It looks like this merge may have caused existing benchmark tests to break...
http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/104/artifact/results/KAFKA-2527/2015-10-13--001/report.html

The producer performance service is timing out in several benchmark tests now.

File: tests/kafkatest/services/performance/producer_performance.py [hunk 11]
```diff
@@ -38,12 +41,14 @@ def __init__(self, context, num_nodes, kafka, security_protocol, topic, num_reco
...
 
     def _worker(self, idx, node):
         args = self.args.copy()
-        args.update({'bootstrap_servers': self.kafka.bootstrap_servers()})
-        cmd = "/opt/kafka/bin/kafka-run-class.sh org.apache.kafka.clients.tools.ProducerPerformance "\
-              "%(topic)s %(num_records)d %(record_size)d %(throughput)d bootstrap.servers=%(bootstrap_servers)s" % args
+        args.update({'bootstrap_servers': self.kafka.bootstrap_servers(), 'jmx_port': self.jmx_port, 'client_id': self.client_id})
+        cmd = "JMX_PORT=%(jmx_port)d /opt/kafka/bin/kafka-run-class.sh org.apache.kafka.clients.tools.ProducerPerformance " \
+              "%(topic)s %(num_records)d %(record_size)d %(throughput)d bootstrap.servers=%(bootstrap_servers)s client.id=%(client_id)s" \
+              " | tee /mnt/producer-performance.log" % args
```
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: @lindong28 
I tried this a bit locally, and realized the command now has two pipes to `tee` (see line 58 as well)

When I drop the pipe to tee here on line 51 and keep the one below, the producer runs as expected.

**Comment by granders** [comment]
@gwenshap 
Confirming that failed benchmark tests pass once again locally. Running the full suite again on aws: http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/105/console

**Comment by gwenshap** [comment]
LGTM.

**Comment by lindong28** [comment]
@granders @ewencp @gwenshap Thanks for your review!

**Comment by b-goyal** [comment]
@granders Can you take a look at this quota system test?
